include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+            params.flags = static_cast<IImageView<image_t>::E_CREATE_FLAGS>(0);
+            params.format = _format;
+            params.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+            params.subresourceRange.baseArrayLayer = 0u;
+            params.subresourceRange.layerCount = image->getCreationParameters().arrayLayers;
+            params.subresourceRange.baseMipLevel = 0u;
+            params.subresourceRange.levelCount = image->getCreationParameters().mipLevels;
+            params.image = core::smart_refctd_ptr<image_t>(image);
+            params.viewType = asset::IImageView<image_t>::ET_2D_ARRAY;
+
+            return m_viewsCache.insert({_format,createView_internal(std::move(params))}).first->second;
+        }
+        auto getPageTableLayers() const
+        {
+            return std::make_pair(m_assignedPageTableLayers.begin(),m_assignedPageTableLayers.end());
+        }
+        void addPageTableLayer(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first,rng.second,_layer);
+            m_assignedPageTableLayers.insert(it, _layer);
+        }
+        void removeLayerAssignment(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first, rng.second, _layer);
+            if (it!=rng.second && *it==_layer)
+                m_assignedPageTableLayers.erase(it);
+        }
+
+        inline uint32_t physPgOffset_x(SPhysPgOffset _offset) const { return _offset.addr & m_addr_xMask; }
+        inline uint32_t physPgOffset_y(SPhysPgOffset _offset) const { return (_offset.addr >> (m_addr_layerShift>>1)) & m_addr_xMask; }
+        inline uint32_t physPgOffset_layer(SPhysPgOffset _offset) const { return (_offset.addr & 0xffffu)>>m_addr_layerShift; }
+        inline bool physPgOffset_valid(SPhysPgOffset _offset) const { return (_offset.addr&0xffffu) != SPhysPgOffset::invalid_addr; }
+        inline bool physPgOffset_hasMipTailAddr(SPhysPgOffset _offset) const { return physPgOffset_valid(physPgOffset_mipTailAddr(_offset)); }
+        inline SPhysPgOffset physPgOffset_mipTailAddr(SPhysPgOffset _offset) const { return _offset.addr>>16; }
+        //! @returns texel-wise offset of physical page
+        core::vector3du32_SIMD pageCoords(SPhysPgOffset _txoffset, uint32_t _pageSz, uint32_t _padding) const
+        {
+            core::vector3du32_SIMD coords(physPgOffset_x(_txoffset), physPgOffset_y(_txoffset), 0u);
+            coords *= (_pageSz + 2u*_padding);
+            coords += _padding;
+            coords.z = physPgOffset_layer(_txoffset);
+            return coords;
+        }
+
+        core::smart_refctd_ptr<image_t> image;
+        using phys_pg_addr_alctr_t = core::PoolAddressAllocator<uint32_t>;
+        uint8_t* m_alctrReservedSpace = nullptr;
+        phys_pg_addr_alctr_t tileAlctr;
+        core::vector<uint16_t> m_assignedPageTableLayers;
```

Comment:
I'd assign viewformats (not format classes) to layers, not layers to format classes, this is going to backfire

Reply:
See https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files#r418001409

Reply:
we can actually get rid of `m_assignedPageTableLayers` and `m_freePageTableLayerIDs`

See https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r418120642